lib/elm-doc-server.ts

```diff
@@ -165,6 +166,21 @@ function completeApplication(manifestPath: string, manifest: Manifest): Manifest
   if (!("license" in manifest)) {
     manifest.license = "Fair";
   }
+  if (!("exposed-modules" in manifest) &&
+    "source-directories" in manifest &&
+    manifest["source-directories"].length > 0) {
+
+    const sourceDirs: string[] = manifest["source-directories"];
+    manifest["exposed-modules"] =
+      sourceDirs
+        .map(srcDir => {
+          // get the module names from the relative elm file paths
+          return glob.sync(`**/*.elm`, { cwd: path.resolve(dir, srcDir) }).
+            map(str => str.slice(0, str.length - 4).replace(/\//g, '.'))
+        })
+        .reduce((a, b) => a.concat(b)) //flatten the array
+        .sort();
+  }
```

Comment:
This is tricky and not documented enough in the current source code, but this would break the forked/vendored packages use case.

See: https://github.com/dmy/elm-doc-preview/blob/7bbd89e407cb4c96e64432b63fc2dec0e823cf9f/lib/elm-doc-server.ts#L306
and from the documentation:
> Without an elm-application.json file, elm-doc-preview will show an application as my/application 1.0.0 and will report an error about missing exposed-modules unless some are eventually found in forked or local packages included in the application source-directories.

and

> Exposed modules contain by default those found in forked and local packages (see next section). Setting the field does not remove those modules from the list.

So if a forked package is included in the `source-directories` and its `elm.json` has been kept in the directory above (basically a `git clone` has been made, and the `src` sub-directory of the package added to `source-directories`), we use the `exposed-modules` from its manifest.

For an example, you can check out the source of version 3.0.4, at this time the `project-metadata-utils` was forked and included in the `elm-doc-preview` frontend source directories. If you start `elm-doc-preview` with it, you can see that the `exposed-modules` from the `exposed-modules` are added to the list of the application exposed modules.

The logic should likely be shared with the one from `buildApplicationDocs`, maybe by extracting it in a new function. Some comments should be also added, sorry for that.

Also this is a symptom that the application has become complex enough to require test cases in order to prevent regressions. The `--output` option added recently could allow to have the different use cases tested before a release and on commits/PRs using CI. This could include testing:
* a package
* an application without exposed modules
* an application with exposed modules
* an application with forked packages
* combinations for the 3 previous cases
* running `elm-doc-preview` from the directory with `elm.json`
* running it from another directory with the optional path argument (it can be a directory or a `elm.json`)
* etc.

Consequently I believe we should really add those before adding more use cases. I have not thought about what test framework we could use, do you have an idea? Something that would allow to run `elm-doc-preview -o` on input directories, and compare the resulting `docs.json` with the expected one.